File: frontend/lib/features/library/models/book_comment.dart
// frontend/lib/features/library/models/book_comment.dart

class BookComment {
  final int id;
  final String content;
  final int bookId;
  final int userId;
  final String userName;
  final String? userAvatar;
  final int? parentId;
  final bool isEdited;
  final int likesCount;
  final bool isLiked;  // ✅ Ajouter ce champ
  final DateTime createdAt;
  final DateTime? updatedAt;
  final List<BookComment> replies;

  BookComment({
    required this.id,
    required this.content,
    this.bookId = 0,
    required this.userId,
    required this.userName,
    this.userAvatar,
    this.parentId,
    this.isEdited = false,
    this.likesCount = 0,
    this.isLiked = false,  // ✅ Ajouter avec valeur par défaut
    required this.createdAt,
    this.updatedAt,
    this.replies = const [],
  });

  factory BookComment.fromJson(Map<String, dynamic> json) {
    return BookComment(
      id: json['id'] ?? 0,
      content: json['content'] ?? '',
      bookId: json['book_id'] ?? 0,
      userId: json['user_id'] ?? 0,
      userName: json['user_name'] ?? 'Anonyme',
      userAvatar: json['user_avatar'],
      parentId: json['parent_id'],
      isEdited: json['is_edited'] ?? false,
      likesCount: json['likes_count'] ?? 0,
      isLiked: json['is_liked'] ?? false,  // ✅ Ajouter
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : DateTime.now(),
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at']) 
          : null,
      replies: (json['replies'] as List?)
          ?.map((r) => BookComment.fromJson(r))
          .toList() ?? [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'content': content,
      'parent_id': parentId,
      'is_liked': isLiked,  // ✅ Ajouter
    };
  }

  BookComment copyWith({
    int? id,
    String? content,
    int? bookId,
    int? userId,
    String? userName,
    String? userAvatar,
    int? parentId,
    bool? isEdited,
    int? likesCount,
    bool? isLiked,  // ✅ Ajouter
    DateTime? createdAt,
    DateTime? updatedAt,
    List<BookComment>? replies,
  }) {
    return BookComment(
      id: id ?? this.id,
      content: content ?? this.content,
      bookId: bookId ?? this.bookId,
      userId: userId ?? this.userId,
      userName: userName ?? this.userName,
      userAvatar: userAvatar ?? this.userAvatar,
      parentId: parentId ?? this.parentId,
      isEdited: isEdited ?? this.isEdited,
      likesCount: likesCount ?? this.likesCount,
      isLiked: isLiked ?? this.isLiked,  // ✅ Ajouter
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      replies: replies ?? this.replies,
    );
  }
}

File: backend/app/models/book_comment.py
# backend/app/models/book_comment.py

from sqlalchemy import Column, Integer, Text, DateTime, ForeignKey, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..core.database import Base


class BookComment(Base):
    __tablename__ = "book_comments"
    __table_args__ = {'extend_existing': True}

    id = Column(Integer, primary_key=True, index=True)
    content = Column(Text, nullable=False)
    book_id = Column(Integer, ForeignKey("books.id"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    parent_id = Column(Integer, ForeignKey("book_comments.id"), nullable=True)
    is_edited = Column(Boolean, default=False)
    likes_count = Column(Integer, default=0)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, onupdate=func.now())

    # ✅ Relations
    book = relationship("Book", back_populates="comments", foreign_keys=[book_id])
    user = relationship("User", back_populates="book_comments", foreign_keys=[user_id])
    
    # ✅ Self-relationship pour les réponses
    parent = relationship("BookComment", remote_side=[id], backref="replies")
    
    # ✅ Relation avec les likes
    likes = relationship("CommentLike", back_populates="comment", cascade="all, delete-orphan")

    def to_dict(self, include_replies=True, user_id=None):
        result = {
            "id": self.id,
            "content": self.content,
            "book_id": self.book_id,
            "user_id": self.user_id,
            "user_name": self.user.full_name if self.user else "Anonyme",
            "user_avatar": self.user.avatar_url if self.user else None,
            "parent_id": self.parent_id,
            "is_edited": self.is_edited,
            "likes_count": self.likes_count,
            "is_liked": any(like.user_id == user_id for like in self.likes) if user_id else False,  # ✅ Vérifier si l'utilisateur a liké
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }
        
        if include_replies and hasattr(self, 'replies'):
            result["replies"] = [r.to_dict(include_replies=False, user_id=user_id) for r in self.replies if r.id != self.id]
        else:
            result["replies"] = []
            
        return result